### PR TITLE
EVG-16477 Move s3.go to pail

### PR DESCRIPTION
**Pull request by malikchaya2**
[EVG-16477](https://jira.mongodb.org/browse/EVG-16477)
The functions in [s3.go|https://github.com/evergreen-ci/evergreen/blob/main/thirdparty/s3.go ] should live here. 
This repo is a more appropriate place for it as it'll help us reduce the amount of miscellaneous third-party code in evergreen. 

Once this is merge, I'll create the corresponding evergreen PR.

### CODE REVIEW

File: s3_bucket.go [hunk 1]
```diff
@@ -1323,3 +1325,64 @@ func (s *s3ArchiveBucket) Pull(ctx context.Context, opts SyncOptions) error {
 
 	return nil
 }
+
+// PresignExpireTime sets the amount of time the link is live before expiring.
+const PresignExpireTime = 24 * time.Hour
```
Comment (julianedwards) [inline s3_bucket.go:1330]: Why not make this configurable?

File: s3_bucket.go [hunk 2]
```diff
@@ -1323,3 +1325,64 @@ func (s *s3ArchiveBucket) Pull(ctx context.Context, opts SyncOptions) error {
...
+
+// RequestParams holds all the parameters needed to sign a url or fetch headObject.
+type RequestParams struct {
+	Bucket    string `json:"bucket"`
+	FileKey   string `json:"fileKey"`
```
Comment (julianedwards) [inline s3_bucket.go:1335]: Nit: Can we use snake case for the json tags? And for consistency change `FileKey` to `Key`.

File: s3_bucket.go [hunk 3]
```diff
@@ -1323,3 +1325,64 @@ func (s *s3ArchiveBucket) Pull(ctx context.Context, opts SyncOptions) error {
...
+	AwsKey    string `json:"awsKey"`
+	AwsSecret string `json:"awsSecret"`
+	Region    string `json:"region"`
+}
+
+// PreSign returns a presigned url that expires in 24 hours.
```
Comment (julianedwards) [inline s3_bucket.go:1341]: Nit: `URL`

File: s3_bucket.go [hunk 4]
```diff
@@ -1323,3 +1325,64 @@ func (s *s3ArchiveBucket) Pull(ctx context.Context, opts SyncOptions) error {
...
+	req, _ := svc.GetObjectRequest(&s3.GetObjectInput{
+		Bucket: aws.String(r.Bucket),
+		Key:    aws.String(r.FileKey),
+	})
+
+	urlStr, err := req.Presign(PresignExpireTime)
+	return urlStr, err
+}
+
+// GetHeadObject fetches the metadata of an s3 object.
```
Comment (julianedwards) [inline s3_bucket.go:1364]: Nit: `S3`

File: s3_bucket.go [hunk 5]
```diff
@@ -1323,3 +1325,64 @@ func (s *s3ArchiveBucket) Pull(ctx context.Context, opts SyncOptions) error {
 
 	return nil
 }
+
+// PresignExpireTime sets the amount of time the link is live before expiring.
+const PresignExpireTime = 24 * time.Hour
+
+// RequestParams holds all the parameters needed to sign a url or fetch headObject.
```
Comment (julianedwards) [inline s3_bucket.go:1332]: `URL` and `fetch S3 object metadata`.

File: s3_bucket.go [hunk 6]
```diff
@@ -1323,3 +1325,64 @@ func (s *s3ArchiveBucket) Pull(ctx context.Context, opts SyncOptions) error {
...
+type RequestParams struct {
```
Comment (julianedwards) [inline s3_bucket.go:1333]: Can we give this a more descriptive name? This package has a lot of implementations of the `Bucket` interface and  in consequence a lot of options structs.

File: s3_bucket.go [hunk 7]
```diff
@@ -1323,3 +1325,64 @@ func (s *s3ArchiveBucket) Pull(ctx context.Context, opts SyncOptions) error {
...
+func PreSign(r RequestParams) (string, error) {
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(endpoints.UsEast1RegionID),
+		Credentials: credentials.NewStaticCredentialsFromCreds(credentials.Value{
+			AccessKeyID:     r.AwsKey,
+			SecretAccessKey: r.AwsSecret,
+		}),
+	})
+	if err != nil {
+		return "", err
```
Comment (julianedwards) [inline s3_bucket.go:1351]: We should probably wrap these errors.

**Review by julianedwards — CHANGES_REQUESTED** [review]

File: s3_bucket.go (hunk 1 above)
Comment (malikchaya2) [reply s3_bucket.go:1330]: Since I'm only moving this from evergreen to pail, it would be out of scope for this ticket.

File: s3_bucket.go (hunk 2 above)
Comment (malikchaya2) [reply s3_bucket.go:1335]: Since this is already being used in prod, I don't think it's a good idea to change it now

File: s3_bucket.go (hunk 3 above)
Comment (malikchaya2) [reply s3_bucket.go:1341]: done

File: s3_bucket.go (hunk 5 above)
Comment (malikchaya2) [reply s3_bucket.go:1332]: done, thank you

File: s3_bucket.go (hunk 7 above)
Comment (malikchaya2) [reply s3_bucket.go:1351]: done

File: s3_bucket.go (hunk 6 above)
Comment (malikchaya2) [reply s3_bucket.go:1333]: I changed it to PreSignRequestParams. Does that work?

Comment (julianedwards) [reply s3_bucket.go]: Thanks!

**Review by julianedwards — APPROVED** [review]

**Pull request merged**